<?xml version="1.0" encoding="UTF-8" standalone="yes" ?>
<!DOCTYPE bugzilla SYSTEM "https://bugs.ecmascript.org/page.cgi?id=bugzilla.dtd">

<bugzilla version="4.4.4+"
          urlbase="https://bugs.ecmascript.org/"
          
          maintainer="dherman@mozilla.com"
>

    <bug>
          <bug_id>3781</bug_id>
          
          <creation_ts>2015-02-05 14:37:00 -0800</creation_ts>
          <short_desc>13.6.4.13 ForIn/OfBodyEvaluation: Perform IteratorClose in 5.a - 5.g ?</short_desc>
          <delta_ts>2015-02-12 12:17:35 -0800</delta_ts>
          <reporter_accessible>1</reporter_accessible>
          <cclist_accessible>1</cclist_accessible>
          <classification_id>1</classification_id>
          <classification>Unclassified</classification>
          <product>Draft for 6th Edition</product>
          <component>technical issue</component>
          <version>Rev 32: February 2, 2015 Draft</version>
          <rep_platform>All</rep_platform>
          <op_sys>All</op_sys>
          <bug_status>RESOLVED</bug_status>
          <resolution>FIXED</resolution>
          
          
          <bug_file_loc></bug_file_loc>
          <status_whiteboard></status_whiteboard>
          <keywords></keywords>
          <priority>Normal</priority>
          <bug_severity>normal</bug_severity>
          <target_milestone>---</target_milestone>
          
          
          <everconfirmed>1</everconfirmed>
          <reporter name="André Bargull">andrebargull</reporter>
          <assigned_to name="Allen Wirfs-Brock">allen</assigned_to>
          <cc>marc.nieper+bugzilla</cc>
          

      

      

      

          <comment_sort_order>oldest_to_newest</comment_sort_order>  
          <long_desc isprivate="0" >
    <commentid>12152</commentid>
    <comment_count>0</comment_count>
    <who name="André Bargull">andrebargull</who>
    <bug_when>2015-02-05 14:37:36 -0800</bug_when>
    <thetext>13.6.4.13 Runtime Semantics: ForIn/OfBodyEvaluation

The newly added IteratorClose calls in destructuring forms, collections etc. call IteratorClose even if the iterator itself throws an exception. Contrary to that 13.6.4.13 only calls IteratorClose when an abrupt completion occurs in step 5.h-i or in 5.k.

Are these differences in the semantics intentional?</thetext>
  </long_desc><long_desc isprivate="0" >
    <commentid>12284</commentid>
    <comment_count>1</comment_count>
    <who name="Allen Wirfs-Brock">allen</who>
    <bug_when>2015-02-10 12:01:00 -0800</bug_when>
    <thetext>fixed in rev33 editor&apos;s draft

I was being overly aggressive in classing IteratorClose.

IteratorClose should not be called if the iterator has normally reached its end or for exceptions that originated from the iterator.

In particular we shouldn&apos;t call the IteratorClose for exceptions that are received via IteratorStep IteratorNext, IteratorConplete, IteratorValue

I fixed Array destructuring and the built-ins to conform to this rule.</thetext>
  </long_desc><long_desc isprivate="0" >
    <commentid>12395</commentid>
    <comment_count>2</comment_count>
    <who name="Allen Wirfs-Brock">allen</who>
    <bug_when>2015-02-12 12:17:35 -0800</bug_when>
    <thetext>fixed in rev33</thetext>
  </long_desc>
      
      

    </bug>

</bugzilla>